scripts/operations/system_recovery/rapid_rebuild.sh

```diff
@@ -0,0 +1,156 @@
+rm -rf /etc/kubernetes/*
+
+# Remove files in /etc/ceph from m001
+rm -rf /etc/ceph/*
+
+# Unmount /etc/cray/upgrade/csm on m001
+umount /etc/cray/upgrade/csm --force
+
+# Remove the all master and worker NCNs from the kubernetes cluster
+pdsh -b -w "$(grep -oP '(ncn-m\w+|ncn-w\w+)' /etc/hosts | sort -u | tr -t '\n' ',')" 'kubeadm reset -f'
+
+# Remove the kube-ipvs0 link -- https://github.com/k3s-io/k3s/issues/5643
+pdsh -b -w "$(grep -oP '(ncn-m\w+|ncn-w\w+)' /etc/hosts | sort -u | tr -t '\n' ',')" 'ip link delete kube-ipvs0'
+
+# Now that k8s is shutdown, start dnsmasq and setup the squash fs links
+systemctl start dnsmasq.service
+
+# Set network config for PIT/dnsmasq on m001
+sed -i 's/^NETCONFIG_DNS_POLICY=.*/NETCONFIG_DNS_POLICY="auto"/' /etc/sysconfig/network/config
+sed -i -E 's/NETCONFIG_DNS_FORWARDER="\w+"/NETCONFIG_DNS_FORWARDER="dnsmasq"/' /etc/sysconfig/network/config
+netconfig update -f
+
+# Setup links for NCN boot images and check DHCP
+/root/bin/set-sqfs-links.sh
+
+# Fix links for worker NCN images
+echo "Updating boot artifact links for NCN worker nodes..."
+pushd /var/www || err_exit "Failed changing directory to /var/www"
+for WDIR in $(ls -d ncn-w*); do
+  echo "Updating boot artifact links for $WDIR..."
+  pushd $WDIR || err_exit "Failed changing directory to $(pwd)/$WDIR"
+  # Modify links to point to worker images
+  for art in kernel initrd.img.xz rootfs; do
+    new_target=$(ls -l $art | awk '{ print $NF }' | sed 's#/data/k8s/#/data/worker/#') || err_exit "Error parsing symbolic link: $(pwd)/$art"
+    [[ -n $new_target ]] || err_exit "Error finding target of symbolic link $(pwd)/$art"
+    [[ -f $new_target ]] || err_exit "File does not exist: $(pwd)/$new_target"
+    ln -snf $new_target $art || err_exit "Command failed: ln -snf $new_target $art"
+  done
+  popd || err_exit "Failed changing directory back to /var/www"
+done
+popd || err_exit "Failed returning to previous directory"
+
+# Edit /etc/dnsmasq.d/statics.conf on m001 for pit nexus
+DNSFILE="/etc/dnsmasq.d/statics.conf"
+NMNIP=$(grep ncn-m001.nmn /etc/hosts | awk '{print $1}')
+pitnexus="host-record=registry,registry.local,packages,packages.local,${NMNIP} # pit nexus"
+sed -i "s/.*registry.*/${pitnexus}/" ${DNSFILE}
+systemctl restart dnsmasq.service
+
+# Edit m001 /etc/hosts to comment out ipv6 entry for ncn-m001 (.e.g. ::1 drax-ncn-m001.local drax-ncn-m001 ncn-m001 x3000c0s1b0n0)
+sed -i '/^::1 '"${SYSTEM_NAME}"'/s/^/#/' /etc/hosts
+
+# Exit if dnsmasq service is not running
+systemctl is-active --quiet dnsmasq.service \
+  && echo "Pit disk dnsmasq service is running" || err_exit "Pit disk dnsmasq service is not running"
+
+export mtoken='ncn-m(?!001)\w+-mgmt'
+export stoken='ncn-s\w+-mgmt'
+export wtoken='ncn-w\w+-mgmt'
+USERNAME=$(whoami)
+export USERNAME
+
+# Power off the NCNs (excluding ncn-m001)
+grep -oP "($mtoken|$stoken|$wtoken)" /etc/dnsmasq.d/statics.conf | sort -u | xargs -t -i ipmitool -I lanplus -U $USERNAME -E -H {} power off
```

Comment:
The same bit:
```
grep -oP "($mtoken|$stoken|$wtoken)" /etc/dnsmasq.d/statics.conf | sort -u
```
is executed 5 times.  It could be executed once and saved in a variable, and then just reuse the variable
```
ncn_list="$(grep -oP "($mtoken|$stoken|$wtoken)" /etc/dnsmasq.d/statics.conf | sort -u)"
echo "$ncn_list" | xargs -t -i ipmitool -I lanplus -U $USERNAME -E -H {} power off
```

Reply:
changed - thanks good idea!